File: rust/src/llm_router.rs
use crate::llm_api::{LlmRequest, LlmResponse, ToolCall};
use reqwest::Client;
use std::error::Error;

pub struct LlmRouter {
    pub provider: String,
    pub api_key: String,
    pub base_url: String,
    client: Client,
}

impl LlmRouter {
    pub fn new(provider: &str, api_key: &str) -> Self {
        let base_url = match provider {
            "openai" => "https://api.openai.com/v1/chat/completions",
            "anthropic" => "https://api.anthropic.com/v1/messages",
            _ => "https://api.openai.com/v1/chat/completions", // Default fallback
        }
        .to_string();

        Self {
            provider: provider.to_string(),
            api_key: api_key.to_string(),
            base_url,
            client: Client::new(),
        }
    }

    pub async fn send_request(&self, request: &LlmRequest) -> Result<LlmResponse, Box<dyn Error>> {
        let mut builder = self.client.post(&self.base_url);

        // Map provider-specific headers
        if self.provider == "anthropic" {
            builder = builder
                .header("x-api-key", &self.api_key)
                .header("anthropic-version", "2023-06-01")
                .header("content-type", "application/json");
        } else {
            builder = builder.bearer_auth(&self.api_key);
        }

        let response = builder.json(request).send().await?;

        // In a full implementation, we would handle provider-specific response parsing here.
        // For now, we assume the response matches the LlmResponse schema for compilation.
        let json_resp = response.json::<LlmResponse>().await?;
        Ok(json_resp)
    }

    pub async fn stream_response<F>(&self, _request: &LlmRequest, _on_chunk: F) -> Result<(), Box<dyn Error>>
    where
        F: Fn(String),
    {
        // Placeholder for SSE parsing logic
        Ok(())
    }

    pub fn handle_tool_call(&self, _tool_call: &ToolCall) -> String {
        // Placeholder for local function routing
        "{}".to_string()
    }
}

File: typescript/src/llmRouter.ts
import { LlmRequest, LlmResponse, ToolCall } from "./llmApi";

export class LlmRouter {
  public provider: string;
  public apiKey: string;
  public baseUrl: string;

  constructor(provider: string, apiKey: string) {
    this.provider = provider;
    this.apiKey = apiKey;

    if (provider.toLowerCase() === "anthropic") {
      this.baseUrl = "https://api.anthropic.com/v1/messages";
    } else {
      this.baseUrl = "https://api.openai.com/v1/chat/completions";
    }
  }

  public async sendRequest(request: LlmRequest, signal?: AbortSignal): Promise<LlmResponse> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json"
    };

    if (this.provider.toLowerCase() === "anthropic") {
      headers["x-api-key"] = this.apiKey;
      headers["anthropic-version"] = "2023-06-01";
    } else {
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }

    const response = await fetch(this.baseUrl, {
      method: "POST",
      headers,
      body: JSON.stringify(request),
      signal
    });

    if (!response.ok) {
      throw new Error(`API request failed with status: ${response.status}`);
    }

    // In a full implementation, provider-specific parsing happens here.
    return response.json() as Promise<LlmResponse>;
  }

  public streamResponse(request: LlmRequest, onChunk: (chunk: string) => void): void {
    // Placeholder for SSE parsing logic
  }

  public handleToolCall(toolCall: ToolCall): string {
    // Placeholder for local function routing
    return "{}";
  }
}
